File: 03/1.py
#!/bin/python3

from functools import reduce
from re import finditer

with open("input.txt") as f:
    lines = [line.strip() for line in f.readlines()]

print(
    reduce(
        lambda acc, e: acc
        + reduce(
            lambda acc, x: acc + int(x.group(0)),
            filter(
                lambda num: len(
                    list(
                        filter(
                            lambda l: len(
                                l[
                                    max(num.start() - 1, 0) : min(num.end() + 1, len(l))
                                ].strip(".0123456789")
                            ),
                            lines[max(e[0] - 1, 0) : min(e[0] + 2, len(lines))],
                        )
                    )
                ),
                finditer("\\d+", e[1]),
            ),
            0,
        ),
        enumerate(lines),
        0,
    )
)

File: 03/1.exs
# {:ok, text} = File.read("test.txt")
{:ok, text} = File.read("input.txt")

lines = text |> String.split("\n")

for {line, i} <- lines |> Enum.with_index, reduce: 0 do acc ->
  all_nums = Regex.scan(~r"\d+", line, return: :index) |> Enum.map(&hd &1)
  valid_nums = all_nums |> Enum.filter( fn {s, len} ->
      ln_slc = lines |> Enum.slice(max(i - 1, 0)..i+1)
      map_ln_slc = ln_slc |> Enum.map(&String.slice(&1, max(s-1, 0)..s+len)) |> Enum.map(&Regex.replace(~r"[\.0-9]", &1, ""))
      filtered = map_ln_slc |> Enum.filter(&(String.length(&1) > 0))
      filtered |> Enum.count > 0
    end)
  num_sum = valid_nums |> Enum.map(fn {s, e} -> String.slice(line, s, e) |> String.to_integer end) |> Enum.sum
  acc + num_sum
end |> IO.puts
